File: ios/Shared/Latency/LatencyStats.swift
import Foundation

/// Pure helpers for latency analysis: percentiles, throughput, and CDF points for distribution overlays.
enum LatencyStats {
    static func percentile(_ sorted: [Double], _ p: Double) -> Double {
        guard !sorted.isEmpty else { return 0 }
        return sorted[min(sorted.count - 1, Int(p * Double(sorted.count)))]
    }

    struct Summary {
        let p50: Double
        let p95: Double
        let p99: Double
        let max: Double
        let throughput: Double   // calls/sec
    }

    static func summary(_ samples: [Double]) -> Summary {
        let sorted = samples.sorted()
        let totalMs = samples.reduce(0, +)
        let throughput = totalMs > 0 ? Double(samples.count) / (totalMs / 1000) : 0
        return Summary(
            p50: percentile(sorted, 0.50),
            p95: percentile(sorted, 0.95),
            p99: percentile(sorted, 0.99),
            max: sorted.last ?? 0,
            throughput: throughput)
    }

    /// Cumulative-distribution points (value, fraction in [0,1]) for overlaying distributions.
    static func cdf(_ samples: [Double]) -> [(value: Double, fraction: Double)] {
        let sorted = samples.sorted()
        let n = Double(sorted.count)
        guard n > 0 else { return [] }
        return sorted.enumerated().map { (value: $0.element, fraction: Double($0.offset + 1) / n) }
    }

    /// "Commit to one unit" (facelift spec §5): sub-millisecond values read as µs — never a bare
    /// `0.11 ms` — so FFI's magnitude stays legible next to HTTP/SQLite. At or above 1 ms, plain ms with
    /// 2 decimals. One value, one unit; never both in the same string. `String(format:)` uses the C
    /// locale (period decimal) regardless of the device locale — the en-US pin Android needs is implicit
    /// here, so a comma-decimal device can't print "1,00 ms" (parity break). Mirrors Android `formatLatencyMs`.
    static func formatLatencyMs(_ ms: Double) -> String {
        ms < 1.0
            ? String(format: "%.1f µs", ms * 1_000.0)
            : String(format: "%.2f ms", ms)
    }

    /// Indices of `samples` (in original order) whose value exceeds the `p`th percentile — the tail worth
    /// annotating on a jitter/sequence chart (facelift spec §5). Empty below 5 samples. NOTE: `percentile`
    /// is nearest-rank (`sorted[Int(p*n)]`), so at the default p=0.99 the threshold IS the max for any
    /// n ≤ 100 (index n-1) — zero outliers for small n. A non-empty result needs n well above 100 (the
    /// real caller, JitterView, uses 400). Mirrors Android `outlierIndices`.
    static func outlierIndices(_ samples: [Double], p: Double = 0.99) -> [Int] {
        guard samples.count >= 5 else { return [] }
        let threshold = percentile(samples.sorted(), p)
        return samples.indices.filter { samples[$0] > threshold }
    }

    /// X-positions where a (x, y) series drops by more than `dropFraction` of its own peak from the
    /// previous point — a GC collection give-back on a heap-over-time series (facelift spec §5). Pure, so
    /// the detector is unit-tested; the UI supplies (index-or-seconds, MB) pairs. Mirrors Android
    /// `collectionEventXs`.
    static func collectionEventXs(_ points: [(x: Double, y: Double)], dropFraction: Double = 0.10) -> [Double] {
        guard points.count >= 2 else { return [] }
        let peak = max(points.map(\.y).max() ?? 0, 1e-9)
        let threshold = peak * dropFraction
        return (1..<points.count)
            .filter { points[$0 - 1].y - points[$0].y > threshold }
            .map { points[$0].x }
    }
}

File: android/app/src/main/kotlin/io/dotnetnativeinterop/feature/LatencyStats.kt
package io.dotnetnativeinterop.feature

/** One latency histogram bucket: its midpoint (ms) and how many samples fell in it.
 *  Mirrors the iOS `LatencyBin` (Shared/Latency/DistributionView.swift). */
public data class LatencyBin(val midpoint: Double, val count: Int)

/**
 * Pure helpers for latency analysis: percentiles, throughput, CDF points, and histogram binning.
 * Byte-for-byte mirror of the iOS `LatencyStats` enum (Shared/Latency/LatencyStats.swift) plus
 * `LatencyBin.make`. No Android dependencies, so it is exercised directly by JVM unit tests.
 */
public object LatencyStats {

    /** Percentile of an already-sorted ascending series (0 for an empty series). */
    public fun percentile(sorted: List<Double>, p: Double): Double {
        if (sorted.isEmpty()) return 0.0
        val index = minOf(sorted.size - 1, (p * sorted.size).toInt())
        return sorted[index]
    }

    public data class Summary(
        val p50: Double,
        val p95: Double,
        val p99: Double,
        val max: Double,
        /** calls/sec, derived from the summed client round-trip time. */
        val throughput: Double,
    )

    public fun summary(samples: List<Double>): Summary {
        val sorted = samples.sorted()
        val totalMs = samples.sum()
        val throughput = if (totalMs > 0.0) samples.size / (totalMs / 1000.0) else 0.0
        return Summary(
            p50 = percentile(sorted, 0.50),
            p95 = percentile(sorted, 0.95),
            p99 = percentile(sorted, 0.99),
            max = sorted.lastOrNull() ?: 0.0,
            throughput = throughput,
        )
    }

    /** Cumulative-distribution point: a value (ms) and the fraction of samples at or below it. */
    public data class CdfPoint(val value: Double, val fraction: Double)

    /** Cumulative-distribution points for overlaying distributions. */
    public fun cdf(samples: List<Double>): List<CdfPoint> {
        val sorted = samples.sorted()
        val n = sorted.size
        if (n == 0) return emptyList()
        return sorted.mapIndexed { i, value -> CdfPoint(value, (i + 1).toDouble() / n) }
    }

    /** Buckets [samples] into [bucketCount] equal-width bins for a histogram. */
    public fun bins(samples: List<Double>, bucketCount: Int): List<LatencyBin> {
        if (samples.isEmpty() || bucketCount < 1) return emptyList()
        val low = samples.min()
        val high = samples.max()
        if (high <= low) return listOf(LatencyBin(low, samples.size))
        val width = (high - low) / bucketCount
        val counts = IntArray(bucketCount)
        for (sample in samples) {
            val index = minOf(bucketCount - 1, ((sample - low) / width).toInt())
            counts[index]++
        }
        return counts.mapIndexed { index, count ->
            LatencyBin(midpoint = low + ((index + 0.5) * width), count = count)
        }
    }

    /**
     * "Commit to one unit" (facelift spec §5): sub-millisecond values read as µs — never a bare
     * `0.11 ms` — so FFI's magnitude stays legible next to HTTP/SQLite. At or above 1 ms, plain ms with
     * 2 decimals. One value, one unit; never both in the same string.
     */
    public fun formatLatencyMs(ms: Double): String =
        if (ms < 1.0) "%.1f µs".format(ms * 1_000.0) else "%.2f ms".format(ms)

    /**
     * Indices of [samples] (in original order) whose value exceeds the [p]th percentile — the tail
     * worth annotating on a jitter/sequence chart (facelift spec §5: "annotate outliers"). Empty below
     * 5 samples, where a percentile isn't meaningful.
     */
    public fun outlierIndices(samples: List<Double>, p: Double = 0.99): List<Int> {
        if (samples.size < 5) return emptyList()
        val threshold = percentile(samples.sorted(), p)
        return samples.indices.filter { samples[it] > threshold }
    }

    /**
     * X-positions where a (x, y) series drops by more than [dropFraction] of its own peak from the
     * previous point — a GC collection give-back on a heap/committed-over-time series (facelift spec
     * §5: "annotate ... GC events"). Pure so the detector is unit-tested; the UI supplies engine
     * BenchmarkPoints as (seconds, MB) pairs.
     */
    public fun collectionEventXs(points: List<Pair<Double, Double>>, dropFraction: Double = 0.10): List<Double> {
        if (points.size < 2) return emptyList()
        val peak = points.maxOf { it.second }.coerceAtLeast(1e-9)
        val threshold = peak * dropFraction
        return (1 until points.size)
            .filter { points[it - 1].second - points[it].second > threshold }
            .map { points[it].first }
    }
}
